plumbing/transport/http/dumb.go

```diff
@@ -0,0 +1,432 @@
+	"github.com/go-git/go-git/v5/plumbing/format/packfile"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/utils/ioutil"
+)
+
+func (s *HTTPSession) fetchDumb(ctx context.Context, req *transport.FetchRequest) error {
+	if req.Depth != 0 {
+		return fmt.Errorf("dumb http protocol does not support shallow capabilities")
+	}
+
+	fsi, ok := s.st.(interface {
+		Filesystem() billy.Filesystem
+	})
+	if !ok {
+		return fmt.Errorf("dumb http protocol requires a filesystem")
+	}
+
+	repoFs := fsi.Filesystem()
+	r := newFetchWalker(s, ctx, repoFs)
+	if err := r.process(); err != nil {
+		return err
+	}
+
+	if err := r.fetch(); err != nil {
+		return fmt.Errorf("error fetching objects: %w", err)
+	}
+
+	return nil
+}
+
+// fetchWalker implements the Dumb protocol for fetching objects.
+type fetchWalker struct {
+	*HTTPSession
+	ctx     context.Context
+	fs      billy.Filesystem
+	queue   []plumbing.Hash
+	packIdx map[plumbing.Hash]string
+}
+
+func newFetchWalker(s *HTTPSession, ctx context.Context, fs billy.Filesystem) *fetchWalker {
+	walker := new(fetchWalker)
+	walker.HTTPSession = s
+	walker.ctx = ctx
+	walker.fs = fs
+	walker.queue = make([]plumbing.Hash, 0)
+	walker.packIdx = make(map[plumbing.Hash]string)
+	return walker
+}
+
+func (r *fetchWalker) getInfoPacks() ([]string, error) {
+	url := fmt.Sprintf("%s/objects/info/packs", r.ep.String())
+	req, err := http.NewRequestWithContext(r.ctx, "GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	applyHeaders(req, "", r.ep, r.auth, "", false)
+	res, err := doRequest(r.client, req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Body.Close()
+	switch res.StatusCode {
+	case http.StatusOK:
+		// continue
+	case http.StatusNotFound:
+		return nil, transport.ErrRepositoryNotFound
+	default:
+		return nil, fmt.Errorf("unexpected status code: %d", res.StatusCode)
+	}
+
+	var packs []string
+	s := bufio.NewScanner(res.Body)
+	for s.Scan() {
+		line := s.Text()
+		hash := strings.TrimPrefix(line, "P pack-")
+		hash = strings.TrimSuffix(hash, ".pack")
+		packs = append(packs, hash)
+	}
+
+	return packs, s.Err()
+}
+
+// downloadFile downloads a file from the server and saves it to the filesystem.
+func (r *fetchWalker) downloadFile(path string) error {
+	url := fmt.Sprintf("%s/%s", r.ep.String(), path)
+	req, err := http.NewRequestWithContext(r.ctx, "GET", url, nil)
+	if err != nil {
+		return err
+	}
+
+	applyHeaders(req, "", r.ep, r.auth, "", false)
+	res, err := doRequest(r.client, req)
+	if err != nil {
+		return err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected status code: %d", res.StatusCode)
+	}
+
+	f, err := r.fs.TempFile(filepath.Dir(path), filepath.Base(path)+".temp")
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(f, res.Body); err != nil {
+		_ = f.Close()
+		return err
+	}
+
+	if err := f.Close(); err != nil {
+		return err
+	}
```

Comment:
As `res.Body` [must](https://github.com/golang/go/blob/858a0e9dfd10ac94a0b9de4429749f0cb99e8cb8/src/net/http/response.go#L61-L64) always be read to completion, maybe something like:

```suggestion
	copy := func(w io.Writer) error {
		if _, err := io.Copy(w, res.Body); err != nil {
			return err
		}

		if err := f.Close(); err != nil {
			return err
		}
	}
	
	f, err := r.fs.TempFile(filepath.Dir(path), filepath.Base(path)+".temp")
	if err != nil {
		copy(io.Discard)	
		return err
	}

	copy(f)	
```